From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Thu, 31 Jul 2025 19:13:13 +0900
Subject: [PATCH] optimize get chunk


diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java
index 36ce2be30478d734d8326eeeb004ba7dc61e0642..92c77c8fe79738dff149687278cbc385ba631a20 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java
@@ -78,6 +78,7 @@ public final class ChunkHolderManager {
     final ChunkUnloadQueue unloadQueue;
 
     private final ConcurrentLong2ReferenceChainedHashTable<NewChunkHolder> chunkHolders = ConcurrentLong2ReferenceChainedHashTable.createWithCapacity(16384, 0.25f);
+    private final it.unimi.dsi.fastutil.longs.Long2ReferenceOpenHashMap<NewChunkHolder> chunkHoldersNoSync = new it.unimi.dsi.fastutil.longs.Long2ReferenceOpenHashMap<>(1024, 0.5F); // Leaf - optimize get chunk
     private final ServerLevel world;
     private final ChunkTaskScheduler taskScheduler;
     private long currentTick;
@@ -374,6 +375,8 @@ public final class ChunkHolderManager {
                 if (current == null) {
                     // must create
                     current = ChunkHolderManager.this.createChunkHolder(key);
+                    ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(ChunkHolderManager.this.world, CoordinateUtils.getChunkX(key), CoordinateUtils.getChunkZ(key), "Cannot update chunk status asynchronously"); // Leaf - optimize get chunk
+                    ChunkHolderManager.this.chunkHoldersNoSync.put(key, current); // Leaf - optimize get chunk
                     ChunkHolderManager.this.chunkHolders.put(key, current);
                     current.updateTicketLevel(newLevel);
                 } else {
@@ -923,11 +926,24 @@ public final class ChunkHolderManager {
     }
 
     public NewChunkHolder getChunkHolder(final int chunkX, final int chunkZ) {
-        return this.chunkHolders.get(CoordinateUtils.getChunkKey(chunkX, chunkZ));
+        // Leaf start - optimize get chunk
+        final long position = CoordinateUtils.getChunkKey(chunkX, chunkZ);
+        if (ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(world, chunkX, chunkZ)) {
+            return this.chunkHoldersNoSync.get(position);
+        } else {
+            return this.chunkHolders.get(position);
+        }
+        // Leaf end - optimize get chunk
     }
 
     public NewChunkHolder getChunkHolder(final long position) {
-        return this.chunkHolders.get(position);
+        // Leaf start - optimize get chunk
+        if (ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(world, CoordinateUtils.getChunkX(position), CoordinateUtils.getChunkZ(position))) {
+            return this.chunkHoldersNoSync.get(position);
+        } else {
+            return this.chunkHolders.get(position);
+        }
+        // Leaf end - optimize get chunk
     }
 
     public void raisePriority(final int x, final int z, final Priority priority) {
@@ -986,6 +1002,8 @@ public final class ChunkHolderManager {
         }
 
         current = this.createChunkHolder(position);
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(ChunkHolderManager.this.world, CoordinateUtils.getChunkX(position), CoordinateUtils.getChunkZ(position), "Cannot update chunk status asynchronously"); // Leaf - optimize get chunk
+        this.chunkHoldersNoSync.put(position, current); // Leaf - optimize get chunk
         this.chunkHolders.put(position, current);
 
 
@@ -1158,7 +1176,12 @@ public final class ChunkHolderManager {
         holder.onUnload();
         this.autoSaveQueue.remove(holder);
         PlatformHooks.get().onChunkHolderDelete(this.world, holder.vanillaChunkHolder);
-        this.chunkHolders.remove(CoordinateUtils.getChunkKey(holder.chunkX, holder.chunkZ));
+        // Leaf start - optimize get chunk
+        long position = CoordinateUtils.getChunkKey(holder.chunkX, holder.chunkZ);
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(ChunkHolderManager.this.world, CoordinateUtils.getChunkX(position), CoordinateUtils.getChunkZ(position), "Cannot update chunk status asynchronously");
+        this.chunkHoldersNoSync.remove(position);
+        this.chunkHolders.remove(position);
+        // Leaf end - optimize get chunk
     }
 
     // note: never call while inside the chunk system, this will absolutely break everything
diff --git a/net/minecraft/server/level/ServerChunkCache.java b/net/minecraft/server/level/ServerChunkCache.java
index ba0f66b5089e252e5d532cb4b94c6246a65d69d8..d8d064f1e8cf01a363566438c0b1cab92303a463 100644
--- a/net/minecraft/server/level/ServerChunkCache.java
+++ b/net/minecraft/server/level/ServerChunkCache.java
@@ -77,6 +77,7 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
     private volatile NaturalSpawner.SpawnState lastSpawnState; // Leaf - optimize mob spawning
     // Paper start
     public final ca.spottedleaf.concurrentutil.map.ConcurrentLong2ReferenceChainedHashTable<net.minecraft.world.level.chunk.LevelChunk> fullChunks = new ca.spottedleaf.concurrentutil.map.ConcurrentLong2ReferenceChainedHashTable<>();
+    public final it.unimi.dsi.fastutil.longs.Long2ReferenceOpenHashMap<net.minecraft.world.level.chunk.LevelChunk> fullChunksNonSync = new it.unimi.dsi.fastutil.longs.Long2ReferenceOpenHashMap<>(1024, 0.5F); // Leaf - optimize get chunk
     public int getFullChunksCount() {
         return this.fullChunks.size();
     }
@@ -87,16 +88,28 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
     @Override
     public final void moonrise$setFullChunk(final int chunkX, final int chunkZ, final LevelChunk chunk) {
         final long key = ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(chunkX, chunkZ);
+        // Leaf start - optimize get chunk
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.level, chunkX, chunkZ, "Cannot update full chunk status asynchronously");
         if (chunk == null) {
+            this.fullChunksNonSync.remove(key);
             this.fullChunks.remove(key);
         } else {
+            this.fullChunksNonSync.put(key,chunk);
             this.fullChunks.put(key, chunk);
         }
+        // Leaf end - optimize get chunk
     }
 
     @Override
     public final LevelChunk moonrise$getFullChunkIfLoaded(final int chunkX, final int chunkZ) {
-        return this.fullChunks.get(ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(chunkX, chunkZ));
+        // Leaf start - optimize get chunk
+        long key = ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(chunkX, chunkZ);
+        if (ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this.level, chunkX, chunkZ)) {
+            return this.fullChunksNonSync.get(key);
+        } else {
+            return this.fullChunks.get(key);
+        }
+        // Leaf end - optimize get chunk
     }
 
     private ChunkAccess syncLoad(final int chunkX, final int chunkZ, final ChunkStatus toStatus) {
@@ -282,7 +295,14 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
 
     @Nullable
     public LevelChunk getChunkAtIfLoadedImmediately(int x, int z) {
-        return this.fullChunks.get(ChunkPos.asLong(x, z));
+        // Leaf start - optimize get chunk
+        long key = ChunkPos.asLong(x, z);
+        if (ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this.level, x, z)) {
+            return this.fullChunksNonSync.get(key);
+        } else {
+            return this.fullChunks.get(key);
+        }
+        // Leaf end - optimize get chunk
     }
     // Paper end
 
@@ -317,7 +337,8 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
     public ChunkAccess getChunk(int x, int z, ChunkStatus chunkStatus, boolean requireChunk) {
         // Paper start - rewrite chunk system
         if (chunkStatus == ChunkStatus.FULL) {
-            final LevelChunk ret = this.fullChunks.get(ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(x, z));
+            long key = ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(x, z); // Leaf - optimize get chunk
+            final LevelChunk ret = (ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this.level, x, z)) ? this.fullChunksNonSync.get(key) : this.fullChunks.get(key); // Leaf - optimize get chunk
 
             if (ret != null) {
                 return ret;
@@ -334,7 +355,8 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
     @Override
     public LevelChunk getChunkNow(int chunkX, int chunkZ) {
         // Paper start - rewrite chunk system
-        final LevelChunk ret = this.fullChunks.get(ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(chunkX, chunkZ));
+        long key = ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(chunkX, chunkZ); // Leaf - optimize get chunk
+        final LevelChunk ret = (ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this.level, chunkX, chunkZ)) ? this.fullChunksNonSync.get(key) : this.fullChunks.get(key); // Leaf - optimize get chunk
         if (!ca.spottedleaf.moonrise.common.PlatformHooks.get().hasCurrentlyLoadingChunk()) {
             return ret;
         }
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 2a1c89bb6712c97ad48b9f64c26f36afe8c811e4..c18614dbb86e843752fba35ee0b6ac422b5bea54 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -1397,12 +1397,17 @@ public abstract class Level implements LevelAccessor, UUIDLookup<Entity>, AutoCl
             }
         }
         // CraftBukkit end
-        if (this.isOutsideBuildHeight(pos)) {
+        // Leaf start - optimize get chunk
+        int x = pos.getX();
+        int y = pos.getY();
+        int z = pos.getZ();
+        if (this.isOutsideBuildHeight(y)) {
             return Blocks.VOID_AIR.defaultBlockState();
         } else {
-            ChunkAccess chunk = this.getChunk(pos.getX() >> 4, pos.getZ() >> 4, ChunkStatus.FULL, true); // Paper - manually inline to reduce hops and avoid unnecessary null check to reduce total byte code size, this should never return null and if it does we will see it the next line but the real stack trace will matter in the chunk engine
-            return chunk.getBlockState(pos);
+            ChunkAccess chunk = this.getChunk(x >> 4, z >> 4, ChunkStatus.FULL, true); // Paper - manually inline to reduce hops and avoid unnecessary null check to reduce total byte code size, this should never return null and if it does we will see it the next line but the real stack trace will matter in the chunk engine
+            return chunk.getBlockState(x, y, z);
         }
+        // Leaf end - optimize get chunk
     }
 
     @Override
@@ -1522,7 +1527,7 @@ public abstract class Level implements LevelAccessor, UUIDLookup<Entity>, AutoCl
             // Spigot end
             if (tickingBlockEntity.isRemoved()) {
                 ((org.dreeam.leaf.util.list.BlockEntityTickersList) this.blockEntityTickers).markAsRemoved(this.tileTickPosition); // toRemove.add(tickingBlockEntity); // SparklyPaper - optimize block entity removals // Paper - Fix MC-117075; use removeAll
-            } else if (runsNormally && this.shouldTickBlocksAt(tickingBlockEntity.getPos())) {
+            } else if (runsNormally /*&& this.shouldTickBlocksAt(tickingBlockEntity.getPos())*/) { // Leaf - optimize get chunk - checked in LevelChunk#isTicking
                 tickingBlockEntity.tick();
                 // Paper start - rewrite chunk system
                 // Leaf start - SparklyPaper - parallel world ticking (only run mid-tick at the end of each tick / fixes concurrency bugs related to executeMidTickTasks) - do not bother with condition work / make configurable
diff --git a/net/minecraft/world/level/chunk/LevelChunk.java b/net/minecraft/world/level/chunk/LevelChunk.java
index c3ecdd80efad340c9fa0ea6913db9a3362b53ad4..17be5a6c1fc0a8a6c1c77c624f181cf5eddb4d70 100644
--- a/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/net/minecraft/world/level/chunk/LevelChunk.java
@@ -316,12 +316,18 @@ public class LevelChunk extends ChunkAccess implements ca.spottedleaf.moonrise.p
 
     public BlockState getBlockStateFinal(final int x, final int y, final int z) {
         // Copied and modified from below
-        final int sectionIndex = this.getSectionIndex(y);
-        if (sectionIndex < 0 || sectionIndex >= this.sections.length
-            || this.sections[sectionIndex].nonEmptyBlockCount == 0) {
+        // Leaf start - optimize get chunk
+        final int sectionIndex = (y >> 4) - this.minSection;
+        if (sectionIndex < 0 || sectionIndex >= this.sections.length) {
             return Blocks.AIR.defaultBlockState();
+        } else {
+            LevelChunkSection section = this.sections[sectionIndex];
+            if (section.nonEmptyBlockCount == 0) {
+                return Blocks.AIR.defaultBlockState();
+            }
+            return section.states.get((y & 15) << 8 | (z & 15) << 4 | x & 15);
         }
-        return this.sections[sectionIndex].states.get((y & 15) << 8 | (z & 15) << 4 | x & 15);
+        // Leaf end - optimize get chunk
     }
 
     @Override
diff --git a/net/minecraft/world/level/chunk/PalettedContainer.java b/net/minecraft/world/level/chunk/PalettedContainer.java
index 59d48e7dc0911557c57a7e07f5f9013c010165bd..2400bc237962dd34cba76ebf35c7b28582eeddac 100644
--- a/net/minecraft/world/level/chunk/PalettedContainer.java
+++ b/net/minecraft/world/level/chunk/PalettedContainer.java
@@ -48,6 +48,18 @@ public class PalettedContainer<T> implements PaletteResize<T>, PalettedContainer
     private final PalettedContainer.Strategy strategy;
     //private final ThreadingDetector threadingDetector = new ThreadingDetector("PalettedContainer"); // Paper - unused
 
+    // Leaf start - optimize get chunk
+    private static final java.lang.invoke.VarHandle DATA_HANDLE;
+    static {
+        try {
+            DATA_HANDLE = java.lang.invoke.MethodHandles.lookup()
+                .findVarHandle(PalettedContainer.class, "data", PalettedContainer.Data.class);
+        } catch (ReflectiveOperationException e) {
+            throw new ExceptionInInitializerError(e);
+        }
+    }
+    // Leaf end - optimize get chunk
+
     public void acquire() {
         // this.threadingDetector.checkAndLock(); // Paper - disable this - use proper synchronization
     }
@@ -104,7 +116,7 @@ public class PalettedContainer<T> implements PaletteResize<T>, PalettedContainer
     }
 
     private T readPalette(final PalettedContainer.Data<T> data, final int paletteIdx) {
-        final T[] palette = ((ca.spottedleaf.moonrise.patches.fast_palette.FastPaletteData<T>)(Object)data).moonrise$getPalette();
+        final T[] palette = data.moonrise$getPalette(); // Leaf - optimize get chunk - remove cast
         if (palette == null) {
             return this.readPaletteSlow(data, paletteIdx);
         }
@@ -268,7 +280,7 @@ public class PalettedContainer<T> implements PaletteResize<T>, PalettedContainer
 
     public T get(int index) { // Paper - public
         // Paper start - optimise palette reads
-        final PalettedContainer.Data<T> data = this.data;
+        final PalettedContainer.Data<T> data = (PalettedContainer.Data<T>) DATA_HANDLE.getOpaque(this); // Leaf - optimize get chunk
         return this.readPalette(data, data.storage.get(index));
         // Paper end - optimise palette reads
     }
